packages/editor/src/components/unsaved-changes-warning/index.js

```diff
@@ -45,4 +42,14 @@ class UnsavedChangesWarning extends Component {
 	}
 }
 
-export default UnsavedChangesWarning;
+export default withSelect( ( select ) => ( {
+	// We need to call the selector directly in the listener to avoid race
+	// conditions with `BrowserURL` where `componentDidUpdate` gets the
+	// new value of `isEditedPostDirty` before this component does,
+	// causing this component to incorrectly think a trashed post is still dirty.
+	isDirty: () => {
+		const { __experimentalGetDirtyEntityRecords } = select( 'core' );
+		const dirtyEntityRecords = __experimentalGetDirtyEntityRecords();
+		return dirtyEntityRecords.length > 0;
+	},
+} ) )( UnsavedChangesWarning );
```

Comment:
Nit: There's an opportunity here to refactor the component as a function component + hooks.

Reply:
Yep, David and I have been discussing that! I was recommending doing that in a follow-up PR though.